grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -435,17 +438,78 @@ private static String getResponseBody(List<FieldDescriptor> topLevelFields,
         if (StringUtil.isNullOrEmpty(responseBody)) {
             return null;
         }
-        for (FieldDescriptor fieldDescriptor: topLevelFields) {
+        for (FieldDescriptor fieldDescriptor : topLevelFields) {
             if (fieldDescriptor.getName().equals(responseBody)) {
                 return responseBody;
             }
         }
         return null;
     }
 
+    private static MediaType getMediaTypeFromHttpBody(JsonNode jsonNode) {
+        final String contentType = jsonNode.get("contentType").asText();
+        try {
+            return MediaType.parse(contentType);
+        } catch (IllegalArgumentException e) {
+            logger.warn("Invalid media type in http body content_type {}.", contentType);
+            return MediaType.JSON_UTF_8;
+        }
+    }
+
+    private static Function<HttpData, MediaType> generateMediaTypeDecider(TranscodingSpec spec) {
+        if (HttpBody.getDescriptor().equals(spec.methodDescriptor.getOutputType())) {
+            return httpData -> {
+                try (HttpData data = httpData) {
+                    final byte[] array = data.array();
+                    try {
+                        final JsonNode jsonNode = mapper.readValue(array, JsonNode.class);
```

Comment:
Question) Is it possible to refactor so that only one deserialization is done? We seem to be doing this at `generateResponseBodyConverter` as well.

Reply:
There is a race condition, response body converter first deserializes the JSON, finds the `body` parameter and converts the body to the inner body parameter. So we lose content-type data from the original body.

A way to solve this would be returning body + media_type bundled together by updating the `generateResponseBodyConverter` method to be `generateResponseBodyAndMediaTypeConverter`.

Reply:
I am not entirely sure if it is worth the effort the abstract this out. I feel like readability of the code dropped quite a bit when I have done this. I can try introducing a method that takes `HttpResponse` and emits another `HttpResponse` instead of the two converters. This method can transform headers & body on the go.

Reply:
I have used `AggregatedHttpResponse` because it seems like the response was already aggregated right before we passed it to the converter. LMK if anything looks odd.